uniffi_bindgen/src/lib.rs

```diff
@@ -0,0 +1,267 @@
+//! to process your `.idl` file. This will generate some Rust code to be included in the top-level source
+//! code of your crate. If your IDL file is named `example.idl`, then your build script would call:
+//!
+//! ```text
+//! uniffi_build::generate_scaffolding("./src/example.idl")
+//! ```
+//!
+//! This would output a rust file named `example.uniffi.rs`, ready to be
+//! included into the code of your rust crate like this:
+//!
+//! ```text
+//! include!(concat!(env!("OUT_DIR"), "/example.uniffi.rs"));
+//! ```
+//!
+//! ### 4) Generate foreign language bindings for the library
+//!
+//! The `uniffi-bindgen` utility provides a command-line tool that can produce code to
+//! consume the Rust library in any of several supported languages.
+//! It is done by calling (in kotlin for example):
+//!
+//! ```text
+//! uniffi-bindgen --language kotlin ./src/example.idl
+//! ```
+//!
+//! This will produce a file `example.kt` in the same directory as the .idl file, containing kotlin bindings
+//! to load and use the compiled rust code via its C-compatible FFI.
+//!
+
+const BINDGEN_VERSION: &str = env!("CARGO_PKG_VERSION_MAJOR");
+
+use anyhow::{anyhow, bail, Result};
+use std::convert::TryInto;
+use std::io::prelude::*;
+use std::{
+    collections::HashMap,
+    ffi::OsStr,
+    fs::File,
+    path::{Path, PathBuf},
+    process::Command,
+};
+
+pub mod bindings;
+pub mod interface;
+pub mod scaffolding;
+
+use bindings::TargetLanguage;
+use interface::ComponentInterface;
+use scaffolding::RustScaffolding;
+
+// Generate the infrastructural Rust code for implementing the IDL interface,
+// such as the `extern "C"` function definitions and record data types.
+pub fn generate_component_scaffolding(
+    idl_file: &OsStr,
+    out_dir_override: Option<&OsStr>,
+    manifest_path_override: Option<&OsStr>,
+    format_code: bool,
+) -> Result<()> {
+    let idl_file = PathBuf::from(idl_file)
+        .canonicalize()
+        .map_err(|e| anyhow!("Failed to find idl file: {:?}", e))?;
+    let component = parse_idl(&idl_file)?;
+    ensure_versions_compatibility(&idl_file, manifest_path_override)?;
+    let mut filename = Path::new(&idl_file)
+        .file_stem()
+        .ok_or_else(|| anyhow!("not a file"))?
+        .to_os_string();
+    filename.push(".uniffi.rs");
+    let mut out_dir = get_out_dir(&idl_file, out_dir_override)?;
+    out_dir.push(filename);
+    let mut f =
+        File::create(&out_dir).map_err(|e| anyhow!("Failed to create output file: {:?}", e))?;
+    write!(f, "{}", RustScaffolding::new(&component))
+        .map_err(|e| anyhow!("Failed to write output file: {:?}", e))?;
+    if format_code {
+        Command::new("rustfmt").arg(&out_dir).status()?;
+    }
+    Ok(())
+}
+
+// If the crate for which we are generating bindings for depends on
+// a `uniffi` runtime version that doesn't agree with our own version,
+// the developer of that said crate will be in a world of pain.
+fn ensure_versions_compatibility(
```

Comment:
Added this to ensure the crates compatibility between the bindgen and the runtime crates. This basically compares the major versions and fails if they are different.

Reply:
Great, thank you!

One nit: I think things are changing fast enough (and we're pre-1.0-enough) that we should just do a full version number check here rather than just the major version, at least until we start being deliberate about releasing versions with semver. Put another way: the commit that prepares our 1.0 release should be the one that makes this do a major-version-only comparison.